src/goto-instrument/cover.cpp

```diff
@@ -218,6 +218,299 @@ void collect_operands(const exprt &src, std::vector<exprt> &dest)
+    result.insert(e1);
+  }
+
+  return result;
+}
+
+/*******************************************************************\
+
+Function: is_arithmetic_predicate
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+bool is_arithmetic_predicate(const exprt &src)
+{
+  return (src.id()==ID_lt
+          || src.id()==ID_le
+          || src.id()==ID_equal
+          || src.id()==ID_ge
+          || src.id()==ID_gt
+          || src.id()==ID_notequal);
+}
+
+/*******************************************************************\
+
+Function: replacement_and_conjunction
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: To replace the i-th expr of ''operands'' with each
+          expr inside ''replacement_exprs''.
+
+\*******************************************************************/
+
+std::set<exprt> replacement_and_conjunction(
+  const std::set<exprt> &replacement_exprs,
+  const std::vector<exprt> &operands,
+  const std::size_t i)
+{
+  std::set<exprt> result;
+  for(auto &y : replacement_exprs)
+  {
+    std::vector<exprt> others;
+    for(std::size_t j=0; j<operands.size(); j++)
+      if(i!=j)
+        others.push_back(operands[j]);
+
+    others.push_back(y);
+    exprt c=conjunction(others);
+    result.insert(c);
+  }
+  return result;
+}
+        
+/*******************************************************************\
+
+Function: non_ordered_expr_expansion
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: To expand all non ordered predicates within an expr
+
+\*******************************************************************/
+
+std::set<exprt> non_ordered_expr_expansion(const exprt &src)
+{
+  std::set<exprt> result;
+
+  std::set<exprt> s1, s2;
+
+  if(src.id()!=ID_not||!is_arithmetic_predicate(src.op0()))
+    s1.insert(src);
+  else
+  {
+    auto res=ordered_negation(src.op0());
+    s1.insert(res.begin(), res.end());
+  }
+
+  // expand negations and non-ordered predicates
+  bool changed=true;
+  while(changed)
+  {
+    changed=false;
+    for(auto &x: s1)
+    {
+      std::vector<exprt> operands;
+      collect_operands(x, operands);
+      for(std::size_t i=0; i<operands.size(); i++)
+      {
+        std::set<exprt> res;
+        if(operands[i].id()==ID_not)
+        {
+          exprt no=operands[i].op0();
+          if(is_arithmetic_predicate(no))
+          {
+            res=ordered_negation(no);
+            if(!res.empty()) changed=true;
+          }
+        }
+        else
+        {
+          if(operands[i].id()==ID_le
+             || operands[i].id()==ID_ge
+             || operands[i].id()==ID_notequal)
+          {
+            res=non_ordered_predicate_expansion(operands[i]);
+            if(!res.empty()) changed=true;
+          }
+        }
+        std::set<exprt> co=replacement_and_conjunction(res, operands, i);
+        s2.insert(co.begin(), co.end());
+        if(!res.empty()) break;
+      }
+      if(!changed) s2.insert(x);
+    }
+    if(!changed) break;
```

Comment:
This loop looks like programming with gotos, except that gotos would at least have the possibility of meaningful labels. If at least there were some comments saying what the individual branches try to achieve. Also, the s2.insert(x) might be replicated and placed with the individual cases to make clear that x isn't lost in all those if(...) without an else case.